File: ActiveRealmSample/ActiveRealmSample/Model/Tag.m
//
// Created by Masaki Ando on 2019-06-02.
// Copyright (c) 2019 Hituzi Ando. All rights reserved.
//

#import "Tag.h"

@interface ActiveRealmTag : ARMObject

@property NSString *articleID;
@property NSString *name;

@end

RLM_ARRAY_TYPE(ActiveRealmTag)

@implementation ActiveRealmTag

@end

@implementation Tag

+ (NSDictionary<NSString *, ARMRelationship *> *)definedRelationships {
    return @{
        @"article": [ARMInverseRelationship inverseRelationshipWithClass:NSClassFromString(@"Article")
                                                                    type:ARMInverseRelationshipTypeBelongsTo]
    };
}

+ (BOOL)validateBeforeSaving:(id)obj {
    Tag *tag = obj;

    return tag.articleID.length > 0 && tag.name.length > 0;
}

- (NSString *)description {
    return [self dictionaryWithValuesForKeys:@[
        @"uid",
        @"articleID",
        @"name",
        @"createdAt",
        @"updatedAt"
    ]].description;
}

#pragma mark - property

- (id)article {
    return self.relations[@"article"].object;
}

@end

File: ActiveRealmSwiftSample/ActiveRealmSwiftSample/Model/Tag.swift
//
// Created by Masaki Ando on 2019-06-04.
// Copyright (c) 2019 Hituzi Ando. All rights reserved.
//

import Foundation

import ActiveRealm

class ActiveRealmTag: ARMObject {

    @objc dynamic var articleID = ""
    @objc dynamic var name      = ""
}

class Tag: ARMActiveRealm {

    @objc var articleID = ""
    @objc var name      = ""

    // A relation property. This property is just alias.

    var article: Article {
        return relations["article"]?.object as! Article
    }

    override class func definedRelationships() -> [String: ARMRelationship] {
        return ["article": ARMInverseRelationship(with: Article.self, type: .belongsTo)]
    }

    override class func validateBeforeSaving(_ obj: Any) -> Bool {
        let tag = obj as! Tag

        return !tag.articleID.isEmpty && !tag.name.isEmpty
    }

    override var description: String {
        return asJSONString()
    }
}
